### ISSUE TITLE
Bug: System.IndexOutOfRangeException: Index was outside the bounds of the array.

### ISSUE THREAD

**Issue by Drachenkaetzchen** [rikimaru0345/Ceras#55]
**Describe the bug**

My software tries to deserialize a previously serialized class. During deserialization, the exception `System.IndexOutOfRangeException: Index was outside the bounds of the array.` is raised.

Here's the backtrace:

```
 System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Ceras.SerializerBinary.ReadVarInt(Byte[] bytes, Int32& offset, Int32 bits)
   at Ceras.SerializerBinary.ReadUInt32Bias(Byte[] buffer, Int32& offset, Int32 bias)
   at Ceras.Formatters.ReferenceFormatter`1.Deserialize(Byte[] buffer, Int32& offset, T& value)
   at Ceras.CerasSerializer.Deserialize[T](T& value, Byte[] buffer, Int32& offset, Int32 expectedReadLength)
   at Ceras.CerasSerializer.Deserialize[T](Byte[] buffer)
```

My problem is that I don't know if the saved data has been corrupted or not, and in that case it's quite tricky to create a reproduce code because of quite many classes involved.

I could send over the serialized data file, but I don't know if it's much use without the involved classes.

Basically I'm trying to figure out if it's a bug caused by any change in my code, if it's a bug within Ceras or if the data has been corrupted. Is there any way to figure this out? Note that 200+ serialized files are loaded prior that exception following the same data structure. Also judging from the timestamps there doesn't seem to be a system crash which could have caused a corruption.

The serialized data file is only 2kb in size. How should I proceed with providing additional information?

**How to reproduce the bug**
- None yet, see above

**Platform**
- net4.7+
- Are you using Unity? no

**Comment by rikimaru0345** [comment]
Hi,
one thing that would definitely help is the version tolerance feature. If your data was saved with that active, then the data layout and names are embedded.

If not, then that is a really tricky situation because Ceras's design makes so many tradeoffs in favor of "faster and smaller size".

```
at Ceras.SerializerBinary.ReadUInt32Bias(Byte[] buffer, Int32& offset, Int32 bias)
at Ceras.Formatters.ReferenceFormatter`1.Deserialize(Byte[] buffer, Int32& offset, T& value)
```
If that happened with version tolerance then we have a lead we can investigate. If not, it will be impossible to know at what point the deserialization has gone off the rails.

> Basically I'm trying to figure out if it's a bug caused by any change in my code, if it's a bug within Ceras or if the data has been corrupted. Is there any way to figure this out?

There's no technical way to go about this; unless everything was saved with version tolerance, in which case we could compare the data layout of the binary data we have, and the data layout of the classes/structs that currently exist in the program.

But even if we'd find a difference there, then Ceras would have found it as well and generated a "reader" (instance of SchemaDynamicFormatter) crafted specifically to read the data described by the embedded Schema. Or in other words: if version tolerance was used then Ceras wouldn't (*shouldn't*) have any trouble reading it because that's what version tolerance is all about in the first place...

If you don't use the version tolerance feature then there are lots of things that can break the binary compatibility!
Non version tolerant mode is designed specifically to throw out everything that is not absolutely neccesary (aka: everything except the data). for maximum performance. (That's intended to be used for sending network messages for example, where you'd most likely never save a network message to a file, and all messages are created on one end, sent, and then discarded after having been read.).

In this mode (not using version tolerance) the data layout is assumed to never change, so Ceras derives the layout from the type, field and property names.
That means there are multiple things that affect the data layout, and thus many ways in which it can break if the user ever makes any changes to the classes.

I'll elaborate on that last point in my next post in a few minutes...

**Comment by rikimaru0345** [comment]
Since the data is assumed to never change, all that has to be ensured is that any two computers that run your program with the exact same classes, will end up using the exact same data layout. 
Ceras does that by using type names, property names and field names (and any attributes that are potentially there).

That also means there are lots of (often really subtle) ways to accidentally break the data format.

For example if you make heavy use of polymorphism (derived types, interfaces, ...) then any renamed fields or properties will potentially also cause changes in the data layout of derived types!

Or moving classes into a different folder might cause some 3rd party tool (like ReSharper) to adjust the namespace. But the namespace of any class is also part of the full type name, so that will also count as a breaking change...

>  Note that 200+ serialized files are loaded prior that exception following the same data structure. Also judging from the timestamps there doesn't seem to be a system crash which could have caused a corruption.

Alright, that does indeed sound like the data shouldn't be corrupted.


> The serialized data file is only 2kb in size. How should I proceed with providing additional information?

In case you were using version tolerance, the data format ("Schema") is embedded into the data to allow future versions of your program to read the data. If that is the case, then I could probably manually go through the Schema descriptors manually and *maybe* find out something. It would be a slow and very cumbersome process though. But if its important enough for you I'll see what I can do.

If the data is not saved with version tolerance then there is absolutely no way to interpret the data without also having the same exact classes that were used when the data was serialized. (at most I could extract some strings from it, but only because they appear obvious in a hex editor...).


The best suggestion I have for you would be to revert your program through git to a previous point in time and try to read it with that.
If that doesn't work, then that suggests that there is indeed a bug where Ceras can't read some data it has just created. If you can somehow create a small test with which I can reproduce the problem I can fix it.


In the upcoming Ceras v5 the difference between version tolerant and "fast mode" will grow even further. Both modes will try to improve upon their strengths even more.
The fast mode will become even faster (so even better suited for things like networking, but thus even more brittle/unreliable for long term storage); while VersionTolerance will become even more robust against any kind of changes.

**Comment by Drachenkaetzchen** [comment]
Oof, I've been hunting a red herring. The problem was not with the file I was talking about, but the subsequent file which has a size of 0 bytes - which then of course explains why I receive that particular exception (must have been visually impaired 2 times).

I don't know why the file is 0 bytes, because the save code is pretty simple, like this:

```
var data = ceras.Serialize(obj);
File.WriteAllBytes(dataFile, data);
```

I'm wondering right now: What if `obj` was null? Could that be the cause of returning 0 bytes?

**Comment by rikimaru0345** [comment]
Even serializing `null` should return at least one byte, indicating that the passed reference was in fact null. 

Unless you're implementing some custom `IFormatter` implementation and do some strange things... but I doubt that's what is going on here.

So no, unless you're working with some very strange settings (maybe disabling reference serialization, or a custom formatter) Ceras should never give 0 bytes as a result.

Is it possible that your program was closed while saving a file?

**Comment by Drachenkaetzchen** [comment]
It's unlikely, as the following data files show a timestamp indicating that it wasn't closed.

I think I'll close the issue, it's way too much guesswork for both of us until I can come up with more details what's going on ;)
